tls/s2n_post_handshake.c

```diff
@@ -13,66 +13,172 @@
+     */
+    if (s2n_stuffer_data_available(message) == 0 && s2n_stuffer_data_available(in) >= message_len) {
+        struct s2n_stuffer full_message = { 0 };
+        struct s2n_blob full_message_blob = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&full_message_blob, s2n_stuffer_raw_read(in, message_len), message_len));
+        RESULT_GUARD_POSIX(s2n_stuffer_init(&full_message, &full_message_blob));
+        RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&full_message, message_len));
+        RESULT_GUARD(s2n_post_handshake_process(conn, &full_message, message_type));
+        return S2N_RESULT_OK;
+    }
+
+    /* Skip unknown message types.
+     * If we can't parse a message, there's no reason to actually allocate memory to store it.
+     */
+    if (!s2n_post_handshake_is_known(message_type)) {
+        uint32_t to_skip = MIN(s2n_stuffer_data_available(in), message_len);
+        RESULT_GUARD_POSIX(s2n_stuffer_skip_read(in, to_skip));
+        if (to_skip < message_len) {
+            /* Rewrite header for next fragment */
+            RESULT_GUARD_POSIX(s2n_stuffer_wipe(message));
+            RESULT_GUARD_POSIX(s2n_stuffer_write_uint8(message, message_type));
+            RESULT_GUARD_POSIX(s2n_stuffer_write_uint24(message, message_len - to_skip));
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        } else {
+            return S2N_RESULT_OK;
         }
     }
 
-    return S2N_SUCCESS;
+    /* If insufficient buffer space, resize */
+    if (s2n_stuffer_space_remaining(message) < message_len) {
+        /* We want to avoid servers allocating memory in response to post-handshake messages
+         * to avoid a potential DDOS / resource exhaustion attack.
+         *
+         * Currently, s2n-tls servers only support the KeyUpdate message,
+         * which should never require additional memory to parse.
+         */
+        RESULT_ENSURE(conn->mode == S2N_CLIENT, S2N_ERR_BAD_MESSAGE);
+
+        uint32_t total_size = message_len + TLS_HANDSHAKE_HEADER_LENGTH;;
+        if (message->alloced) {
+            RESULT_GUARD_POSIX(s2n_stuffer_resize(message, total_size));
+        } else {
+            /* Manually convert our static stuffer to a growable stuffer */
+            RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(message, total_size));
+            RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(message, conn->post_handshake.in_bytes, TLS_HANDSHAKE_HEADER_LENGTH));
+            RESULT_GUARD_POSIX(s2n_stuffer_skip_read(message, TLS_HANDSHAKE_HEADER_LENGTH));
+        }
+    }
+
+    /* Try to read the rest of the message */
+    if (s2n_stuffer_data_available(message) < message_len) {
+        uint32_t remaining = message_len - s2n_stuffer_data_available(message);
+        uint32_t to_read = MIN(remaining, s2n_stuffer_data_available(in));
+        RESULT_GUARD_POSIX(s2n_stuffer_copy(in, message, to_read));
+    }
+    RESULT_ENSURE(s2n_stuffer_data_available(message) == message_len, S2N_ERR_IO_BLOCKED);
+
+    /* Finish processing */
```

Comment:
/* The handshake message has been completely re-assembled if we've gotten this far. Now we can process its contents. */